pytorch_lightning/core/lightning.py

```diff
@@ -263,9 +264,13 @@ def _apply_batch_transfer_handler(
         self, batch: Any, device: Optional[torch.device] = None, dataloader_idx: int = 0
     ) -> Any:
         device = device or self.device
-        batch = self.on_before_batch_transfer(batch, dataloader_idx)
-        batch = self.transfer_batch_to_device(batch, device, dataloader_idx)
-        batch = self.on_after_batch_transfer(batch, dataloader_idx)
+        datahook_source = (
+            _DataHookSource(self) if self.trainer is None else self.trainer._data_connector._datahook_source
+        )
+
+        batch = datahook_source.get_hook("on_before_batch_transfer")(batch, dataloader_idx)
+        batch = datahook_source.get_hook("transfer_batch_to_device")(batch, device, dataloader_idx)
+        batch = datahook_source.get_hook("on_after_batch_transfer")(batch, dataloader_idx)
```

Comment:
while the current patch is bad, this doesn't address what we want the fundamental hook order to be here: https://github.com/PyTorchLightning/pytorch-lightning/issues/10498#issuecomment-973228003

Reply:
left a comment here: https://github.com/PyTorchLightning/pytorch-lightning/issues/10498#issuecomment-973852552
I'd say we can prioritize datamodule with a warning.

Reply:
Resolving as I think you've implemented this already